tools/src/main/java/com/example/core/broker/testkit/SubscriberStub.java

```diff
@@ -36,12 +34,15 @@ public SubscriberStub(String topicId, String groupId, ActorRef topicBuffer, Mate
         return new SubscriberStub<>(topicId, groupId, topicBuffer, materializer);
     }
 
+    private <X extends ActorRef, M, M2> M2 registerActor(X ref, M t) {
+        topicBuffer.tell(new SubscribeToBuffer(groupId, ref), ActorRef.noSender());
+        return (M2) t;
+    }
+
     @Override
     public Source<T, ?> atMostOnceSource() {
         Source<T, ActorRef> source = Source.<T>actorRef(1024, OverflowStrategy.fail());
-        Pair<ActorRef, CompletionStage<Done>> pair = source.toMat(Sink.ignore(), Keep.both()).run(materializer);
-        topicBuffer.tell(new SubscribeToBuffer(groupId, pair.first()), ActorRef.noSender());
-        return source;
+        return source.prependMat(Source.empty(), this::registerActor);
     }
```

Comment:
I believe that `atLeastOnce` can now be implemented in terms of `atMostOnceSource`:

```java
    public CompletionStage<Done> atLeastOnce(Flow<T, Done, ?> flow) {
        return atMostOnceSource()
                .via(flow)
                .toMat(Sink.ignore(), Keep.right())
                .run(materializer);
    }
```